### PR TITLE
Cross modal generation

### PR DESCRIPTION
**Pull request by gallilmaimon**
Add support for cross modal generation e.g Text prompt -> Speech continuation and vice versa for interleaving models. This does not yet support arbitrary "mixed" inputs and outputs.

use metric=cm_generation and provide a glob file to *.txt for text prompts and *.wav (for example) for audio prompts, while specifying the prompt modality and continuation modality. 

Example:
python cli/eval.py metric=cm_generate tokeniser=interleaved_hubert_25 model.pretrained_model=<PATH> tokeniser.params.text_tokeniser_path=Qwen/Qwen2.5-0.5B metric.data_path=data/prompt/*.wav vocoder=vocoder_hubert_25

### CODE REVIEW

File: slamkit/metric/cross_modal_generation.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+
+class CrossModalPromptDataset(Dataset):
+    def __init__(self, glob_path, prompt_length=None, prompt_modality='TEXT', sample_rate=16000, num_files=None):
+        super().__init__()
+        if num_files is None:
+            self.data = glob(glob_path, recursive=True)
+        else:
+            paths = iglob(glob_path, recursive=True)
+            self.data = [next(paths) for _ in range(num_files)]
```
Comment (avishaiElmakies) [inline slamkit/metric/cross_modal_generation.py]: i think we can fix the stop iteration bug in advance

File: slamkit/metric/cross_modal_generation.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                w = w[:int(self.prompt_length * self.sample_rate)]
+            return list([GenerationInput.from_tuple(('SPEECH', w))])
+        elif self.prompt_modality == 'TEXT':
+            return list([GenerationInput.from_tuple(('TEXT', CrossModalPromptDataset._load_text(self.data[idx])))])
+        else:
+            raise ValueError(f'Unknown prompt modality {self.prompt_modality}')
+
+
+def collate_fn(batch):
+    return batch
```
Comment (avishaiElmakies) [inline slamkit/metric/cross_modal_generation.py]: why is this needed?

File: slamkit/tokeniser/interleaving_tokeniser.py [hunk 3]
```diff
@@ -251,12 +265,23 @@ def build_prompt(self, wav: torch.Tensor, lens: Optional[torch.Tensor] = None) -
...
-        audio_repr = self.text_tokeniser.decode(tokens)
-        return torch.tensor([int(num) for num in re.findall(r'<Un(\d+)>', audio_repr)])
+    def decode_sample(self, tokens: torch.tensor, output_modality: str = "SPEECH") -> Union[torch.Tensor, str]:
+        # We currently support only single output modality, not interleaved decoding
+        # Note! this could mean that part of the output is ignored
+        ignore_tokens = [self.text_tokeniser.pad_token_id, self.text_tokeniser.bos_token_id, self.text_tokeniser.eos_token_id]
+        ignore_tokens += [self.text_tokeniser.encode(SPEECH_TOKEN)[0], self.text_tokeniser.encode(TEXT_TOKEN)[0]]
+        if output_modality:
+            ignore_tokens += self.get_ignore_tokens(output_modality)
+        ignore_tokens = torch.tensor([i for i in ignore_tokens if i is not None], device=tokens.device)
```
Comment (avishaiElmakies) [inline slamkit/tokeniser/interleaving_tokeniser.py]: why would i be None? is there a better way to do this?

File: slamkit/metric/cross_modal_generation.py (hunk 1 above)
Comment (gallilmaimon) [reply slamkit/metric/cross_modal_generation.py]: True, I will copy your changes until we make this a shared utility

File: slamkit/tokeniser/interleaving_tokeniser.py (hunk 3 above)
Comment (gallilmaimon) [reply slamkit/tokeniser/interleaving_tokeniser.py]: It is none if the special tokens are None which is true for some of the tokenisers. What is wrong with this?

File: slamkit/metric/cross_modal_generation.py (hunk 2 above)
Comment (gallilmaimon) [reply slamkit/metric/cross_modal_generation.py]: To keep list of list structure in batch. I added a comment

File: slamkit/tokeniser/interleaving_tokeniser.py (hunk 3 above)
Comment (avishaiElmakies) [reply slamkit/tokeniser/interleaving_tokeniser.py]: I don't like the loop here. can we maybe make sure that `get_ignore_tokens` doesn't return none? you can check if the tokens exist and if  they do add them to `ignore tokens`

File: slamkit/metric/cross_modal_generation.py (hunk 1 above)
Comment (avishaiElmakies) [reply slamkit/metric/cross_modal_generation.py]: ok

File: slamkit/tokeniser/interleaving_tokeniser.py (hunk 3 above)
Comment (gallilmaimon) [reply slamkit/tokeniser/interleaving_tokeniser.py]: It is not from get_ignore_tokens it is the list at line 207. I can add the loop there if you prefer, not that it really makes any difference

**Review by avishaiElmakies — APPROVED** [review]
